lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandFMA.cpp

```diff
@@ -70,164 +32,325 @@ Value getStructFromValueTable(ArrayRef<Value> vals,
-  Value strideA1 = isARow ? strideAM : strideAK;
-  int aNumPtr = 8;
-  int K = aShapePerCTA[1];
-  int M = aShapePerCTA[0];
-
-  auto shapePerCTATile = getShapePerCTATile(dLayout);
-  auto sizePerThread = getSizePerThread(dLayout);
-
-  Value _0 = i32_val(0);
-
-  Value mContig = i32_val(sizePerThread[order[1]]);
-
-  // threadId in blocked layout
-  auto threadIds = getThreadIds(thread, shapePerCTATile, sizePerThread, order,
-                                rewriter, loc);
-  Value threadIdM = threadIds[0];
-
-  Value offA0 = isARow ? _0 : mul(threadIdM, mContig);
-  Value offA1 = isARow ? mul(threadIdM, mContig) : _0;
-  SmallVector<Value> aOff(aNumPtr);
-  for (int i = 0; i < aNumPtr; ++i) {
-    aOff[i] = add(mul(offA0, strideA0), mul(offA1, strideA1));
+struct DimNumbers {
+  unsigned batch;
+  unsigned k;
+  unsigned nonK;
+};
+
+/// Put elements from Value vec to appropriate indexes in opValues array.
+///
+/// This function maps elements of 3d sub-tensor in linear array.
+/// Axes are arranged in following order from fastest to slowest: [nonKdim,
+/// kDim, bDim]
+void storeValuesInLinearVector(PatternRewriter &rewriter, Location loc,
+                               SmallVector<Value> &opValues, Value vec,
+                               ArrayRef<unsigned> perThreadTileShape,
+                               unsigned kIdx, unsigned nonKIdx, unsigned bIdx,
+                               const DimNumbers &dim, int vecDim,
+                               ArrayRef<unsigned> opOrder) {
+  auto vecTy = cast<VectorType>(vec.getType());
+  auto vectorSize = vecTy.getNumElements();
+  auto elemTy = vecTy.getElementType();
+  for (int elem = 0; elem < vectorSize; ++elem) {
+    unsigned spatialIdx[3] = {};
+    spatialIdx[dim.batch] = bIdx;
+    spatialIdx[dim.k] = kIdx;
+    spatialIdx[dim.nonK] = nonKIdx;
+    spatialIdx[vecDim] += elem;
+
+    unsigned linearIdx = linearize(spatialIdx, perThreadTileShape, opOrder);
+    opValues[linearIdx] = extract_element(elemTy, vec, i32_val(elem));
   }
-  auto elemTy = typeConverter->convertType(aTensorTy.getElementType());
+}
 
-  Type ptrTy = aSmem.base.getType();
-  SmallVector<Value> aPtrs(aNumPtr);
-  for (int i = 0; i < aNumPtr; ++i)
-    aPtrs[i] = gep(ptrTy, elemTy, aSmem.base, aOff[i]);
+void verifyCTALayout(CTALayoutAttr ctaLayout) {
+  auto ctaSplit = ctaLayout.getCTASplitNum();
+  for (auto split : ctaSplit) {
+    if (split != 1)
+      llvm::report_fatal_error("tensors splited in CGA(thread group clusters) "
+                               "are not supported in FMA dot yet.");
+  }
+}
 
-  SmallVector<Value> vas;
+/// Get a linear offset of first element loaded by thread.
+///
+/// In unswizzled case offset of any element computed with formula:
+/// smem.base + first_element_offset + constant_offset.
+///
+/// first_element_offset depends on lane Id and warp Id
+/// constant_offset depends on value number, which is same for all threads.
+/// \returns first_element_offset
```

Comment:
"... is computed with formula"
"constant_offset depends on value index, ..."

Is it assumed that elements for the same thread are contiguous in LDS?

Reply:
No, it only assumes that there are a fixed stride for every value in thread, which is same for every threads.

Consider an example with three threads(threads 0, 3, 32), mfma 32 dot operand A with kwidth=2.

Value number 0 corresponds to matrix elements:
-  [0, 0] for thread 0
-  [3, 0] for thread 3
-  [0, 2] for thread 32

Value number 2 corresponds to elements:
- [0, 4] for thread 0
- [3, 4] for thread 3
- [0, 6] for thread 32

Multidimensional offsets between 2 and 0 element stored in every thread are following:
- thread 0: [0, 4] - [0, 0] = [0, 4]
- thread 3: [3, 4] - [3, 0] = [0, 4]
- thread 32: [0, 6] - [0, 2] = [0, 4]

So in case memory is not swizzled and plain memory offset depends on multidimensional element coordinates lineary, we can split "thread dependant" part of offset and fixed offset for every value.